### DISCUSSION TITLE
How to set MaxAge,MaxBytes,MaxMsgs,MaxMsgs parameters with configuration file

### DISCUSSION THREAD

**Discussion by rdkumavat1** [nats-io/nats-server#5653] [General]
I am trying to set default settings that will be applied on streams, subject and subscriptions on creation without specifying by the end- user.

I am trying to set it with nats server configuration but not successful, Can you please guide me how to do that and which should be ideal methodology to apply it without code..

Thanks

**Comment by Jarema** [comment]
You cannot do it with Server config. That would be problematic for many reasons - inconsistent configs across cluster, behavior after manual change, lack of possibility to error on startup, as R3 streams need quorum for creation, to mention few problems.

**You can achieve that with**
1. use `nats stream add --config config.json`
2. use [NACK](https://github.com/nats-io/nack) if you're running NATS inside k8s.
3. Clients provide idempotent variant for creating streams and consumers, so they can be then safely run on each startup.

**Reply by rdkumavat1** [reply]
Do you have some reference [doc/code sample] that should guide on how to create a wrapper on NATS configuration and APIs..

Thanks

**Reply by Jarema** [reply]
For which scenario?

**Reply by rdkumavat1** [reply]
I am writing a code that is creating additional layer on top of NATS Client library

Thanks

**Reply by Jarema** [reply]
I assume you're using Go client library.

You should use [JetStream API](https://pkg.go.dev/github.com/nats-io/nats.go/jetstream) and its `CreateOrUpdateStream` and `CreateOrUpdateConsumer` functions.

**Comment by yuzhou-nj** [comment]
Use `nats stream info <stream_name> -j > test.json` :
`cat test.json`
```
{
  "config": {
    "name": "test",
    "subjects": [
      "test.\u003e"
    ],
    "retention": "interest",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": 20000000,
    "max_bytes": -1,
    "max_age": 3600000000000,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false,
    "consumer_limits": {}
  },
  "created": "2024-06-12T11:54:11.6604811Z",
  "state": {
    "messages": 0,
    "bytes": 0,
    "first_seq": 500623445,
    "first_ts": "1970-01-01T00:00:00Z",
    "last_seq": 500623444,
    "last_ts": "2024-07-16T12:22:53.6465141Z",
    "consumer_count": 1
  },
  "cluster": {
    "leader": "nats-77"
  },
  "ts": "2024-07-17T00:49:01.6400042Z"
}
```

Modify max_msgs、max_age...

Then run `nats stream edit  --config=test.json <stream_name>`  and press Y 
or `nats stream edit  --config=test.json <stream_name> -f`

**Reply by rdkumavat1** [reply]
thank you, this will work for me
